File: modules/context-native/ios/ImageCompressionProcessor.swift
import CoreGraphics
import CryptoKit
import Darwin
import Foundation
import ImageIO

enum ImageCompressionProcessor {
  static let revision = "1"
  static let maximumOutputPixels = 4_194_304
  static let maximumOutputBytes: Int64 = 52_428_800

  static func inspect(
    fileURL: URL,
    expectedByteCount: Int64,
    expectedSHA256: String,
    cancellation: ImageHashCancellationToken
  ) throws -> [String: Any] {
    try withVerifiedImageSource(
      fileURL: fileURL,
      expectedByteCount: expectedByteCount,
      expectedSHA256: expectedSHA256,
      cancellation: cancellation
    ) { source, properties in
      let dimensions = try orientedDimensions(properties)
      let pixelCount = dimensions.width.multipliedReportingOverflow(by: dimensions.height)
      guard !pixelCount.overflow, pixelCount.partialValue > 0,
            pixelCount.partialValue <= ImagePerceptualHasher.maximumPixelCount else {
        throw ImagePerceptualHashError.resourceLimit
      }
      guard let sourceMediaType = mediaType(source: source) else {
        throw ImagePerceptualHashError.invalidImage
      }
      return [
        "schemaVersion": 1,
        "sourceByteCount": expectedByteCount,
        "sourceSha256": expectedSHA256,
        "sourceMediaType": sourceMediaType,
        "width": dimensions.width,
        "height": dimensions.height,
        "hasAlpha": try imageHasAlpha(source: source, cancellation: cancellation),
        "animated": false,
        "orientationApplied": true,
        "revision": revision,
      ]
    }
  }

  static func compress(
    taskId: String,
    fileURL: URL,
    expectedByteCount: Int64,
    expectedSHA256: String,
    targetWidth: Int,
    targetHeight: Int,
    quality: Double,
    outputMediaType: String,
    preserveAlpha: Bool,
    cancellation: ImageHashCancellationToken,
    beforePublish: (() -> Void)? = nil
  ) throws -> [String: Any] {
    let started = ContinuousClock.now
    let pixelCount = targetWidth.multipliedReportingOverflow(by: targetHeight)
    guard UUID(uuidString: taskId)?.uuidString.lowercased() == taskId,
          targetWidth > 0, targetHeight > 0,
          pixelCount.overflow == false,
          pixelCount.partialValue <= maximumOutputPixels,
          quality >= 0.58, quality <= 1,
          outputMediaType == "image/jpeg" || outputMediaType == "image/png",
          (outputMediaType == "image/png") == preserveAlpha,
          (!preserveAlpha || quality == 1) else {
      throw ImagePerceptualHashError.invalidImage
    }
    let output = try withVerifiedImageSource(
      fileURL: fileURL,
      expectedByteCount: expectedByteCount,
      expectedSHA256: expectedSHA256,
      cancellation: cancellation
    ) { source, properties -> URL in
      let sourceDimensions = try orientedDimensions(properties)
      guard mediaType(source: source) != nil,
            targetWidth <= sourceDimensions.width,
            targetHeight <= sourceDimensions.height,
            try imageHasAlpha(source: source, cancellation: cancellation) == preserveAlpha else {
        throw ImagePerceptualHashError.invalidImage
      }
      let thumbnailOptions: [CFString: Any] = [
        kCGImageSourceCreateThumbnailFromImageAlways: true,
        kCGImageSourceCreateThumbnailWithTransform: true,
        kCGImageSourceThumbnailMaxPixelSize: max(targetWidth, targetHeight),
        kCGImageSourceShouldCacheImmediately: true,
      ]
      guard let thumbnail = CGImageSourceCreateThumbnailAtIndex(
        source,
        0,
        thumbnailOptions as CFDictionary
      ) else {
        try cancellation.check()
        throw ImagePerceptualHashError.invalidImage
      }
      try cancellation.check()
      let outputImage = try render(
        image: thumbnail,
        width: targetWidth,
        height: targetHeight,
        preserveAlpha: preserveAlpha
      )
      let destination = try ImageCompressionTemporaryStore.prepare(taskId: taskId)
      do {
        guard let encoder = CGImageDestinationCreateWithURL(
          destination.partial as CFURL,
          (preserveAlpha ? "public.png" : "public.jpeg") as CFString,
          1,
          nil
        ) else { throw ImagePerceptualHashError.resourceLimit }
        let options: [CFString: Any] = preserveAlpha
          ? [:]
          : [kCGImageDestinationLossyCompressionQuality: quality]
        CGImageDestinationAddImage(encoder, outputImage, options as CFDictionary)
        guard CGImageDestinationFinalize(encoder) else {
          throw ImagePerceptualHashError.invalidImage
        }
        let handle = try FileHandle(forWritingTo: destination.partial)
        do {
          try handle.synchronize()
          try handle.close()
        } catch {
          try? handle.close()
          throw error
        }
        beforePublish?()
        try cancellation.check()
        try FileManager.default.moveItem(
          at: destination.partial,
          to: destination.complete
        )
        ImageCompressionTemporaryStore.register(
          taskId: taskId,
          fileURL: destination.complete
        )
        return destination.complete
      } catch let processingError {
        do {
          try ImageCompressionTemporaryStore.removeUnregistered(
            [destination.partial, destination.complete]
          )
        } catch {
          throw error
        }
        throw processingError
      }
    }
    do {
      try cancellation.check()
      let metadata = try outputMetadata(output, cancellation: cancellation)
      return [
        "schemaVersion": 1,
        "taskId": taskId,
        "sourceSha256": expectedSHA256,
        "temporaryFileUri": output.absoluteString,
        "outputByteCount": metadata.byteCount,
        "outputSha256": metadata.sha256,
        "width": targetWidth,
        "height": targetHeight,
        "mediaType": outputMediaType,
        "quality": quality,
        "alphaPreserved": preserveAlpha,
        "engine": "core-graphics",
        "revision": revision,
        "durationMs": durationMilliseconds(since: started),
      ]
    } catch let processingError {
      guard ImageCompressionTemporaryStore.finish(taskId: taskId) else {
        throw ImagePerceptualHashError.resourceLimit
      }
      throw processingError
    }
  }

  private static func withVerifiedImageSource<T>(
    fileURL: URL,
    expectedByteCount: Int64,
    expectedSHA256: String,
    cancellation: ImageHashCancellationToken,
    operation: (CGImageSource, [CFString: Any]) throws -> T
  ) throws -> T {
    try cancellation.check()
    let snapshot = try ImmutableImageSnapshot.create(
      sourceURL: fileURL,
      expectedByteCount: expectedByteCount,
      expectedSHA256: expectedSHA256,
      cancellation: cancellation
    )
    return try snapshot.withURL { snapshotURL in
      let provider = try CancellableImageDataProvider.create(
        url: snapshotURL,
        byteCount: expectedByteCount,
        cancellation: cancellation
      )
      guard let source = CGImageSourceCreateWithDataProvider(provider, nil),
            ImagePerceptualHasher.acceptsFrameCount(CGImageSourceGetCount(source)),
            let properties = CGImageSourceCopyPropertiesAtIndex(source, 0, nil)
              as? [CFString: Any] else {
        try cancellation.check()
        throw ImagePerceptualHashError.invalidImage
      }
      try cancellation.check()
      return try operation(source, properties)
    }
  }

  private static func orientedDimensions(
    _ properties: [CFString: Any]
  ) throws -> (width: Int, height: Int) {
    guard let width = (properties[kCGImagePropertyPixelWidth] as? NSNumber)?.intValue,
          let height = (properties[kCGImagePropertyPixelHeight] as? NSNumber)?.intValue,
          width > 0, height > 0 else {
      throw ImagePerceptualHashError.invalidImage
    }
    let orientation = (properties[kCGImagePropertyOrientation] as? NSNumber)?.intValue ?? 1
    return [5, 6, 7, 8].contains(orientation)
      ? (height, width)
      : (width, height)
  }

  private static func imageHasAlpha(
    source: CGImageSource,
    cancellation: ImageHashCancellationToken
  ) throws -> Bool {
    let options: [CFString: Any] = [
      kCGImageSourceCreateThumbnailFromImageAlways: true,
      kCGImageSourceThumbnailMaxPixelSize: 1,
      kCGImageSourceShouldCacheImmediately: true,
    ]
    guard let image = CGImageSourceCreateThumbnailAtIndex(
      source,
      0,
      options as CFDictionary
    ) else {
      try cancellation.check()
      throw ImagePerceptualHashError.invalidImage
    }
    let alpha = image.alphaInfo
    return alpha != .none && alpha != .noneSkipFirst && alpha != .noneSkipLast
  }

  private static func render(
    image: CGImage,
    width: Int,
    height: Int,
    preserveAlpha: Bool
  ) throws -> CGImage {
    let alpha: CGImageAlphaInfo = preserveAlpha ? .premultipliedLast : .noneSkipLast
    guard let context = CGContext(
      data: nil,
      width: width,
      height: height,
      bitsPerComponent: 8,
      bytesPerRow: width * 4,
      space: CGColorSpaceCreateDeviceRGB(),
      bitmapInfo: CGBitmapInfo.byteOrder32Big.rawValue | alpha.rawValue
    ) else { throw ImagePerceptualHashError.resourceLimit }
    if !preserveAlpha {
      context.setFillColor(gray: 1, alpha: 1)
      context.fill(CGRect(x: 0, y: 0, width: width, height: height))
    }
    context.interpolationQuality = .high
    context.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
    guard let output = context.makeImage() else {
      throw ImagePerceptualHashError.resourceLimit
    }
    return output
  }

  private static func mediaType(source: CGImageSource) -> String? {
    switch CGImageSourceGetType(source) as String? {
    case "public.jpeg": return "image/jpeg"
    case "public.png": return "image/png"
    case "com.compuserve.gif": return "image/gif"
    case "com.microsoft.bmp", "public.bmp": return "image/bmp"
    case "org.webmproject.webp": return "image/webp"
    case "public.heic", "public.heif": return "image/heic"
    default: return nil
    }
  }

  static func outputMetadata(
    _ fileURL: URL,
    cancellation: ImageHashCancellationToken,
    readChunk: ((FileHandle) throws -> Data)? = nil
  ) throws -> (byteCount: Int64, sha256: String) {
    let handle = try FileHandle(forReadingFrom: fileURL)
    defer { try? handle.close() }
    var hash = SHA256()
    var byteCount: Int64 = 0
    while true {
      try cancellation.check()
      let data: Data
      if let readChunk {
        data = try readChunk(handle)
      } else {
        data = try handle.read(upToCount: 64 * 1_024) ?? Data()
      }
      if data.isEmpty { break }
      byteCount += Int64(data.count)
      hash.update(data: data)
    }
    guard byteCount > 0, byteCount <= maximumOutputBytes else {
      throw ImagePerceptualHashError.resourceLimit
    }
    try cancellation.check()
    return (byteCount, hash.finalize().map { String(format: "%02x", $0) }.joined())
  }

  private static func durationMilliseconds(
    since started: ContinuousClock.Instant
  ) -> Double {
    let duration = started.duration(to: .now)
    let components = duration.components
    return Double(components.seconds) * 1_000
      + Double(components.attoseconds) / 1_000_000_000_000_000
  }
}

final class ImageCompressionStartupMaintenanceBarrier: @unchecked Sendable {
  private let lock = NSLock()
  private let completion = DispatchGroup()
  private var started = false

  func start(
    queue: DispatchQueue = DispatchQueue.global(qos: .utility),
    operation: @escaping @Sendable () -> Void
  ) {
    lock.lock()
    guard !started else {
      lock.unlock()
      return
    }
    started = true
    completion.enter()
    lock.unlock()
    queue.async { [completion] in
      defer { completion.leave() }
      operation()
    }
  }

  func waitUntilFinished() {
    lock.lock()
    let shouldWait = started
    lock.unlock()
    if shouldWait { completion.wait() }
  }
}

enum ImageCompressionStartupRecoveryReporter {
  static let eventId = "00000000-0000-4000-8000-000000000014"
  private static let operationLock = NSRecursiveLock()
  private static var unpublishedFailureCode: String?

  static func reconcile(
    container: URL,
    failureCode: String?,
    persistRecovery: (URL, String, String) throws -> Void = { container, code, id in
      try RecoveryMetadataEventStore.persistRecovery(
        container: container,
        code: code,
        id: id
      )
    },
    acknowledge: (URL, String) throws -> Bool = { container, id in
      try RecoveryMetadataEventStore.ack(
        container: container,
        folder: "RecoveryEvents",
        id: id
      )
    }
  ) throws {
    operationLock.lock()
    defer { operationLock.unlock() }
    do {
      if let failureCode {
        guard failureCode == ImagePerceptualHashError.cleanupFailure.stableCode else {
          throw ImagePerceptualHashError.cleanupFailure
        }
        try persistRecovery(container, failureCode, eventId)
      } else {
        _ = try acknowledge(container, eventId)
      }
      unpublishedFailureCode = nil
    } catch {
      if let failureCode { unpublishedFailureCode = failureCode }
      throw error
    }
  }

  static func retryPendingPublication(container: URL) throws {
    operationLock.lock()
    defer { operationLock.unlock() }
    if let failureCode = unpublishedFailureCode {
      try reconcile(container: container, failureCode: failureCode)
    }
  }
}

enum ImageCompressionTemporaryStore {
  private static let lock = NSLock()
  private static let maintenanceLock = NSLock()
  private static var outputs: [String: URL] = [:]
  private static var startupFailureCode: String?
  private static let directoryName = "ImageCompression"
  private static let sessionPrefix = UUID().uuidString.lowercased() + "-"

  static func startupMaintenance(
    fileManager: FileManager = .default,
    remover: (URL) throws -> Void = { try FileManager.default.removeItem(at: $0) }
  ) throws {
    maintenanceLock.lock()
    defer { maintenanceLock.unlock() }
    for attempt in 0..<2 {
      do {
        let root = try directory(fileManager: fileManager)
        for child in try fileManager.contentsOfDirectory(
          at: root,
          includingPropertiesForKeys: nil
        ) {
          if child.lastPathComponent.hasPrefix(sessionPrefix) { continue }
          do {
            try remover(child)
          } catch {
            if try !existsNoFollow(child) { continue }
            throw error
          }
        }
        startupFailureCode = nil
        return
      } catch {
        if attempt == 1 {
          startupFailureCode = ImagePerceptualHashError.cleanupFailure.stableCode
          throw ImagePerceptualHashError.cleanupFailure
        }
      }
    }
  }

  @discardableResult
  static func runStartupMaintenance(
    fileManager: FileManager = .default,
    remover: (URL) throws -> Void = { try FileManager.default.removeItem(at: $0) }
  ) -> String? {
    do {
      try startupMaintenance(fileManager: fileManager, remover: remover)
      return nil
    } catch {
      return ImagePerceptualHashError.cleanupFailure.stableCode
    }
  }

  static func currentStartupFailureCode() -> String? {
    maintenanceLock.lock()
    defer { maintenanceLock.unlock() }
    return startupFailureCode
  }

  static func prepare(taskId: String) throws -> (partial: URL, complete: URL) {
    guard UUID(uuidString: taskId)?.uuidString.lowercased() == taskId else {
      throw ImagePerceptualHashError.invalidImage
    }
    // This synchronous fence makes inherited-output cleanup observable before
    // the process can create or publish a new derivative.
    try startupMaintenance()
    let root = try directory()
    let complete = root.appendingPathComponent("\(sessionPrefix)\(taskId).tmp")
    let partial = root.appendingPathComponent("\(sessionPrefix)\(taskId).tmp.partial")
    try removeIfPresent(partial)
    guard try !existsNoFollow(complete) else {
      throw ImagePerceptualHashError.resourceLimit
    }
    return (partial, complete)
  }

  static func register(taskId: String, fileURL: URL) {
    lock.lock()
    outputs[taskId] = fileURL
    lock.unlock()
  }

  static func removeUnregistered(_ fileURLs: [URL]) throws {
    for fileURL in fileURLs {
      try removeIfPresent(fileURL)
    }
  }

  @discardableResult
  static func finish(taskId: String) -> Bool {
    lock.lock()
    let output = outputs[taskId]
    lock.unlock()
    guard let output else { return true }
    let removed: Bool
    do {
      try FileManager.default.removeItem(at: output)
      removed = true
    } catch {
      removed = (try? !existsNoFollow(output)) == true
    }
    guard removed else { return false }
    lock.lock()
    if outputs[taskId] == output { outputs.removeValue(forKey: taskId) }
    lock.unlock()
    return true
  }

  private static func directory(fileManager: FileManager = .default) throws -> URL {
    let value = fileManager.temporaryDirectory.appendingPathComponent(
      directoryName,
      isDirectory: true
    )
    try fileManager.createDirectory(
      at: value,
      withIntermediateDirectories: true,
      attributes: [.posixPermissions: 0o700]
    )
    let metadata = try value.resourceValues(forKeys: [
      .isDirectoryKey,
      .isSymbolicLinkKey,
    ])
    guard metadata.isDirectory == true, metadata.isSymbolicLink != true else {
      throw ImagePerceptualHashError.resourceLimit
    }
    try fileManager.setAttributes(
      [.posixPermissions: 0o700],
      ofItemAtPath: value.path
    )
    return value
  }

  private static func existsNoFollow(_ fileURL: URL) throws -> Bool {
    var metadata = stat()
    if lstat(fileURL.path, &metadata) == 0 { return true }
    if errno == ENOENT { return false }
    throw ImagePerceptualHashError.resourceLimit
  }

  private static func removeIfPresent(_ fileURL: URL) throws {
    if unlink(fileURL.path) == 0 || errno == ENOENT { return }
    throw ImagePerceptualHashError.resourceLimit
  }
}

File: modules/context-native/android/src/main/java/com/aicontextpack/nativebridge/ImageCompressionProcessor.kt
package com.aicontextpack.nativebridge

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Matrix
import android.media.ExifInterface
import android.net.Uri
import android.os.SystemClock
import android.system.Os
import android.system.OsConstants
import android.system.ErrnoException
import java.io.File
import java.io.FileOutputStream
import java.security.MessageDigest
import java.util.UUID

internal object ImageCompressionProcessor {
  const val revision = "1"
  const val maximumOutputPixels = 4_194_304L
  const val maximumOutputBytes = 52_428_800L

  fun inspect(
    context: Context,
    fileUri: String,
    expectedByteCount: Long,
    expectedSha256: String,
    cancellation: ImageHashCancellationToken,
  ): Map<String, Any> {
    cancellation.throwIfCancelled()
    val source = controlledSandboxFile(context, fileUri)
    return ImmutableImageSnapshot.create(
      context,
      source,
      expectedByteCount,
      expectedSha256,
      cancellation,
    ).useDeleting { snapshot ->
      val metadata = inspectSnapshot(snapshot, expectedByteCount, cancellation)
      mapOf(
        "schemaVersion" to 1,
        "sourceByteCount" to expectedByteCount.toDouble(),
        "sourceSha256" to expectedSha256,
        "sourceMediaType" to metadata.mediaType,
        "width" to metadata.orientedWidth,
        "height" to metadata.orientedHeight,
        "hasAlpha" to metadata.hasAlpha,
        "animated" to false,
        "orientationApplied" to true,
        "revision" to revision,
      )
    }
  }

  fun compress(
    context: Context,
    taskId: String,
    fileUri: String,
    expectedByteCount: Long,
    expectedSha256: String,
    targetWidth: Int,
    targetHeight: Int,
    quality: Double,
    outputMediaType: String,
    preserveAlpha: Boolean,
    cancellation: ImageHashCancellationToken,
    beforePublish: (() -> Unit)? = null,
  ): Map<String, Any> {
    val started = SystemClock.elapsedRealtimeNanos()
    val targetPixels = targetWidth.toLong() * targetHeight.toLong()
    if (
      !isCanonicalTaskId(taskId) ||
      targetWidth <= 0 || targetHeight <= 0 ||
      targetPixels !in 1..maximumOutputPixels ||
      !quality.isFinite() || quality !in 0.58..1.0 ||
      outputMediaType !in setOf("image/jpeg", "image/png") ||
      (outputMediaType == "image/png") != preserveAlpha ||
      (preserveAlpha && quality != 1.0)
    ) throw NativeException("PROCESSOR_OUTPUT_INVALID")
    val source = controlledSandboxFile(context, fileUri)
    val output = ImmutableImageSnapshot.create(
      context,
      source,
      expectedByteCount,
      expectedSha256,
      cancellation,
    ).useDeleting { snapshot ->
      val metadata = inspectSnapshot(snapshot, expectedByteCount, cancellation)
      if (
        targetWidth > metadata.orientedWidth || targetHeight > metadata.orientedHeight ||
        preserveAlpha != metadata.hasAlpha
      ) throw NativeException("PROCESSOR_OUTPUT_INVALID")
      val bitmap = decodeBounded(
        snapshot,
        expectedByteCount,
        metadata,
        targetWidth,
        targetHeight,
        cancellation,
      )
      val target = try {
        Bitmap.createBitmap(targetWidth, targetHeight, Bitmap.Config.ARGB_8888)
      } catch (_: OutOfMemoryError) {
        bitmap.recycle()
        throw NativeException("RESOURCE_MEMORY_PRESSURE")
      }
      try {
        val canvas = Canvas(target)
        if (!preserveAlpha) canvas.drawColor(Color.WHITE)
        val matrix = sourceToTargetMatrix(
          bitmap.width,
          bitmap.height,
          targetWidth,
          targetHeight,
          metadata.orientation,
        )
        canvas.drawBitmap(bitmap, matrix, null)
        cancellation.throwIfCancelled()
        writeTemporary(
          context,
          taskId,
          target,
          outputMediaType,
          quality,
          cancellation,
          beforePublish,
        )
      } finally {
        bitmap.recycle()
        target.recycle()
      }
    }
    return try {
      cancellation.throwIfCancelled()
      val digest = digest(output, cancellation)
      mapOf(
        "schemaVersion" to 1,
        "taskId" to taskId,
        "sourceSha256" to expectedSha256,
        "temporaryFileUri" to Uri.fromFile(output).toString(),
        "outputByteCount" to digest.byteCount.toDouble(),
        "outputSha256" to digest.sha256,
        "width" to targetWidth,
        "height" to targetHeight,
        "mediaType" to outputMediaType,
        "quality" to quality,
        "alphaPreserved" to preserveAlpha,
        "engine" to "android-bitmap",
        "revision" to revision,
        "durationMs" to (SystemClock.elapsedRealtimeNanos() - started) / 1_000_000.0,
      )
    } catch (error: Throwable) {
      if (!ImageCompressionTemporaryStore.finish(taskId)) {
        throw NativeException("RESOURCE_MEMORY_PRESSURE")
      }
      throw error
    }
  }

  private data class Metadata(
    val width: Int,
    val height: Int,
    val orientedWidth: Int,
    val orientedHeight: Int,
    val orientation: Int,
    val mediaType: String,
    val hasAlpha: Boolean,
  )

  private fun inspectSnapshot(
    file: File,
    expectedByteCount: Long,
    cancellation: ImageHashCancellationToken,
  ): Metadata {
    cancellation.throwIfCancelled()
    if (ImagePerceptualHasher.violatesSingleFramePolicy(file.path, cancellation)) {
      throw NativeException("PROCESSOR_OUTPUT_INVALID")
    }
    val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
    cancellation.attachDecode(bounds)
    try {
      file.inputStream().use { source ->
        CancellableBoundedInputStream(source, expectedByteCount, cancellation).use { input ->
          BitmapFactory.decodeStream(input, null, bounds)
        }
      }
    } finally {
      cancellation.detachDecode(bounds)
    }
    cancellation.throwIfCancelled()
    val pixelCount = bounds.outWidth.toLong() * bounds.outHeight.toLong()
    if (
      bounds.outWidth <= 0 || bounds.outHeight <= 0 ||
      pixelCount !in 1..ImagePerceptualHasher.maximumPixelCount
    ) throw NativeException("RESOURCE_MEMORY_PRESSURE")
    val mediaType = when (bounds.outMimeType) {
      "image/jpeg", "image/png", "image/gif", "image/webp", "image/heic", "image/heif" ->
        bounds.outMimeType
      "image/bmp", "image/x-ms-bmp" -> "image/bmp"
      else -> throw NativeException("PROCESSOR_OUTPUT_INVALID")
    }
    val orientation = ImagePerceptualHasher.readOrientation(
      file,
      expectedByteCount,
      cancellation,
    )
    val swapsAxes = orientation in setOf(
      ExifInterface.ORIENTATION_TRANSPOSE,
      ExifInterface.ORIENTATION_ROTATE_90,
      ExifInterface.ORIENTATION_TRANSVERSE,
      ExifInterface.ORIENTATION_ROTATE_270,
    )
    val sample = boundedSampleSize(bounds.outWidth, bounds.outHeight, 1, 1)
    val alphaProbe = decodeBitmap(file, expectedByteCount, sample, cancellation)
    val hasAlpha = try { alphaProbe.hasAlpha() } finally { alphaProbe.recycle() }
    return Metadata(
      width = bounds.outWidth,
      height = bounds.outHeight,
      orientedWidth = if (swapsAxes) bounds.outHeight else bounds.outWidth,
      orientedHeight = if (swapsAxes) bounds.outWidth else bounds.outHeight,
      orientation = orientation,
      mediaType = mediaType,
      hasAlpha = hasAlpha,
    )
  }

  private fun decodeBounded(
    file: File,
    expectedByteCount: Long,
    metadata: Metadata,
    targetWidth: Int,
    targetHeight: Int,
    cancellation: ImageHashCancellationToken,
  ): Bitmap {
    val swapsAxes = metadata.orientedWidth != metadata.width
    val rawTargetWidth = if (swapsAxes) targetHeight else targetWidth
    val rawTargetHeight = if (swapsAxes) targetWidth else targetHeight
    val sample = boundedSampleSize(
      metadata.width,
      metadata.height,
      rawTargetWidth,
      rawTargetHeight,
    )
    return decodeBitmap(file, expectedByteCount, sample, cancellation)
  }

  private fun boundedSampleSize(
    width: Int,
    height: Int,
    minimumWidth: Int,
    minimumHeight: Int,
  ): Int {
    var sample = 1
    while (
      width / (sample * 2) >= minimumWidth &&
      height / (sample * 2) >= minimumHeight &&
      ceilDivide(width, sample * 2).toLong() * ceilDivide(height, sample * 2) >= 1
    ) sample *= 2
    while (
      ceilDivide(width, sample).toLong() * ceilDivide(height, sample) > maximumOutputPixels
    ) sample *= 2
    if (ceilDivide(width, sample) < minimumWidth || ceilDivide(height, sample) < minimumHeight) {
      throw NativeException("RESOURCE_MEMORY_PRESSURE")
    }
    return sample
  }

  private fun decodeBitmap(
    file: File,
    expectedByteCount: Long,
    sample: Int,
    cancellation: ImageHashCancellationToken,
  ): Bitmap {
    val options = BitmapFactory.Options().apply {
      inSampleSize = sample
      inScaled = false
      inPreferredConfig = Bitmap.Config.ARGB_8888
    }
    cancellation.attachDecode(options)
    val bitmap = try {
      file.inputStream().use { source ->
        CancellableBoundedInputStream(source, expectedByteCount, cancellation).use { input ->
          BitmapFactory.decodeStream(input, null, options)
        }
      } ?: throw NativeException("PROCESSOR_OUTPUT_INVALID")
    } catch (error: NativeException) {
      throw error
    } catch (_: OutOfMemoryError) {
      throw NativeException("RESOURCE_MEMORY_PRESSURE")
    } catch (_: Throwable) {
      cancellation.throwIfCancelled()
      throw NativeException("PROCESSOR_OUTPUT_INVALID")
    } finally {
      cancellation.detachDecode(options)
    }
    cancellation.throwIfCancelled()
    return bitmap
  }

  private fun sourceToTargetMatrix(
    width: Int,
    height: Int,
    targetWidth: Int,
    targetHeight: Int,
    orientation: Int,
  ): Matrix {
    val source = floatArrayOf(0f, 0f, width.toFloat(), 0f, 0f, height.toFloat())
    val w = targetWidth.toFloat()
    val h = targetHeight.toFloat()
    val destination = when (orientation) {
      ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> floatArrayOf(w, 0f, 0f, 0f, w, h)
      ExifInterface.ORIENTATION_ROTATE_180 -> floatArrayOf(w, h, 0f, h, w, 0f)
      ExifInterface.ORIENTATION_FLIP_VERTICAL -> floatArrayOf(0f, h, w, h, 0f, 0f)
      ExifInterface.ORIENTATION_TRANSPOSE -> floatArrayOf(0f, 0f, 0f, h, w, 0f)
      ExifInterface.ORIENTATION_ROTATE_90 -> floatArrayOf(w, 0f, w, h, 0f, 0f)
      ExifInterface.ORIENTATION_TRANSVERSE -> floatArrayOf(w, h, w, 0f, 0f, h)
      ExifInterface.ORIENTATION_ROTATE_270 -> floatArrayOf(0f, h, 0f, 0f, w, h)
      else -> floatArrayOf(0f, 0f, w, 0f, 0f, h)
    }
    return Matrix().also {
      if (!it.setPolyToPoly(source, 0, destination, 0, 3)) {
        throw NativeException("PROCESSOR_OUTPUT_INVALID")
      }
    }
  }

  private fun writeTemporary(
    context: Context,
    taskId: String,
    bitmap: Bitmap,
    mediaType: String,
    quality: Double,
    cancellation: ImageHashCancellationToken,
    beforePublish: (() -> Unit)?,
  ): File {
    val files = ImageCompressionTemporaryStore.prepare(context, taskId)
    try {
      FileOutputStream(files.partial).use { stream ->
        val format = if (mediaType == "image/png") {
          Bitmap.CompressFormat.PNG
        } else {
          Bitmap.CompressFormat.JPEG
        }
        if (!bitmap.compress(format, (quality * 100).toInt(), stream)) {
          throw NativeException("PROCESSOR_OUTPUT_INVALID")
        }
        stream.fd.sync()
      }
      beforePublish?.invoke()
      cancellation.throwIfCancelled()
      if (!files.partial.renameTo(files.complete)) {
        throw NativeException("RESOURCE_MEMORY_PRESSURE")
      }
      ImageCompressionTemporaryStore.register(taskId, files.complete)
      return files.complete
    } catch (error: Throwable) {
      runCatching {
        ImageCompressionTemporaryStore.removeUnregistered(
          listOf(files.partial, files.complete),
        )
      }.getOrElse { throw it }
      throw error
    }
  }

  private data class Digest(val byteCount: Long, val sha256: String)

  private fun digest(file: File, cancellation: ImageHashCancellationToken): Digest {
    val hash = MessageDigest.getInstance("SHA-256")
    var count = 0L
    val buffer = ByteArray(64 * 1_024)
    file.inputStream().use { input ->
      while (true) {
        cancellation.throwIfCancelled()
        val read = input.read(buffer)
        if (read < 0) break
        if (read == 0) continue
        count += read
        if (count > maximumOutputBytes) throw NativeException("RESOURCE_MEMORY_PRESSURE")
        hash.update(buffer, 0, read)
      }
    }
    if (count <= 0) throw NativeException("PROCESSOR_OUTPUT_INVALID")
    return Digest(count, hash.digest().joinToString("") { "%02x".format(it) })
  }

  private fun ceilDivide(value: Int, divisor: Int): Int =
    (value.toLong() + divisor - 1L).div(divisor).toInt()

  private fun isCanonicalTaskId(value: String): Boolean =
    value == value.lowercase() && runCatching { UUID.fromString(value).toString() == value }
      .getOrDefault(false)
}

internal object ImageCompressionStartupRecoveryReporter {
  const val eventId = "00000000-0000-4000-8000-000000000014"

  fun reconcile(filesDir: File, failureCode: String?) {
    if (failureCode != null) {
      if (failureCode != "PIPELINE_RECOVERY_REQUIRED")
        throw NativeException("PIPELINE_RECOVERY_REQUIRED")
      MetadataEventStore.persistRecovery(filesDir, failureCode, eventId)
      return
    }
    MetadataEventStore.ack(filesDir, "RecoveryEvents", eventId)
  }
}

internal object ImageCompressionTemporaryStore {
  private const val directoryName = "ImageCompression"
  private val sessionPrefix = "${UUID.randomUUID()}-"
  private val outputs = mutableMapOf<String, File>()
  @Volatile private var startupFailureCode: String? = null

  data class Paths(val partial: File, val complete: File)

  @Synchronized
  fun startupMaintenance(
    context: Context,
    remover: (File) -> Boolean = { it.delete() },
  ) {
    repeat(2) { attempt ->
      try {
        val root = directory(context)
        val candidates = root.listFiles()
          ?: throw NativeException("PIPELINE_RECOVERY_REQUIRED")
        candidates.forEach { candidate ->
          if (candidate.name.startsWith(sessionPrefix)) return@forEach
          val stat = Os.lstat(candidate.path)
          if (
            OsConstants.S_ISREG(stat.st_mode) &&
            !remover(candidate) &&
            existsNoFollow(candidate)
          ) throw NativeException("PIPELINE_RECOVERY_REQUIRED")
        }
        startupFailureCode = null
        return
      } catch (_: Exception) {
        if (attempt == 1) {
          startupFailureCode = "PIPELINE_RECOVERY_REQUIRED"
          throw NativeException("PIPELINE_RECOVERY_REQUIRED")
        }
      }
    }
  }

  fun runStartupMaintenance(
    context: Context,
    remover: (File) -> Boolean = { it.delete() },
  ): String? =
    try {
      startupMaintenance(context, remover)
      null
    } catch (_: NativeException) {
      "PIPELINE_RECOVERY_REQUIRED"
    }

  fun currentStartupFailureCode(): String? = startupFailureCode

  fun prepare(context: Context, taskId: String): Paths {
    if (runCatching { UUID.fromString(taskId).toString() == taskId }.getOrDefault(false).not()) {
      throw NativeException("PROCESSOR_OUTPUT_INVALID")
    }
    // Fence inherited-output cleanup before accepting current-process work so
    // cleanup failure is stable and visible through the compression request.
    startupMaintenance(context)
    val root = directory(context)
    val partial = File(root, "$sessionPrefix$taskId.tmp.partial")
    val complete = File(root, "$sessionPrefix$taskId.tmp")
    removeIfPresent(partial)
    if (existsNoFollow(complete)) throw NativeException("RESOURCE_MEMORY_PRESSURE")
    return Paths(partial, complete)
  }

  @Synchronized
  fun register(taskId: String, output: File) {
    outputs[taskId] = output
  }

  fun removeUnregistered(files: List<File>) {
    files.forEach(::removeIfPresent)
  }

  @Synchronized
  fun finish(taskId: String): Boolean {
    val output = outputs[taskId] ?: return true
    val removed = runCatching { removeIfPresent(output) }.isSuccess
    if (removed && outputs[taskId] == output) outputs.remove(taskId)
    return removed
  }

  private fun directory(context: Context): File {
    val root = File(context.cacheDir, directoryName)
    if (!existsNoFollow(root) && !root.mkdirs()) {
      throw NativeException("RESOURCE_MEMORY_PRESSURE")
    }
    val metadata = runCatching { Os.lstat(root.path) }
      .getOrElse { throw NativeException("RESOURCE_MEMORY_PRESSURE") }
    if (!OsConstants.S_ISDIR(metadata.st_mode)) {
      throw NativeException("RESOURCE_MEMORY_PRESSURE")
    }
    runCatching { Os.chmod(root.path, OsConstants.S_IRWXU) }
      .getOrElse { throw NativeException("RESOURCE_MEMORY_PRESSURE") }
    return root
  }

  private fun existsNoFollow(file: File): Boolean = try {
    Os.lstat(file.path)
    true
  } catch (error: ErrnoException) {
    if (error.errno == OsConstants.ENOENT) false
    else throw NativeException("RESOURCE_MEMORY_PRESSURE")
  } catch (_: Exception) {
    throw NativeException("RESOURCE_MEMORY_PRESSURE")
  }

  private fun removeIfPresent(file: File) {
    if (runCatching { file.delete() }.getOrDefault(false)) return
    if (!existsNoFollow(file)) return
    throw NativeException("RESOURCE_MEMORY_PRESSURE")
  }
}
